docs(tip-1077): add expiring nonce time wheel (#5582)
tempoxyz/tempo: tips/tip-1077.md

5. Traffic is assumed to be roughly constant over the 30 valid `valid_before` seconds. A workload
   that intentionally concentrates all transactions into one exact `valid_before` second is still
   accepted probabilistically, but has higher collision pressure.
6. Expiring nonce replay cells are bounded temporary protocol state. They do not pay TIP-1000
   permanent new-slot pricing on first touch and do not mint or consume TIP-1060 storage credits.

If the target TPS assumption is exceeded for long periods, collision frequency and probe-exhausted
rejections increase. A future TIP can raise `EXPIRING_NONCE_BUCKET_CAPACITY` without changing the
replay algorithm.

## Threat Model

Users and relayers are untrusted. They may replay old signed transactions, submit many independent
expiring nonce transactions with the same `valid_before`, or grind transaction contents to seek
cell collisions.

Builders and validators are untrusted for transaction ordering but trusted to execute the canonical
state transition rules. A block containing a replay, an expired expiring nonce transaction, or a
transaction whose probe path is exhausted is invalid.

> Offchain cleanup services are not trusted for replay safety. This TIP does not rely on any MEV,

Reviewer: nit: Not sure this is really relevant for this TIP

particular, 65,536 cells per bucket doubles load to 30.52%, and 32,768 cells per bucket pushes load
above 60%, where probe length and probe-exhaustion risk become much more sensitive to traffic
bursts or adversarial hash selection.

Larger tables reduce collisions but mostly buy marginal latency improvement at the cost of much
larger bounded state. The 16,777,216-cell parameter explored in implementation experiments has
negligible collisions but a 32 GiB slot+value full-table footprint.

## Hardfork Migration

Before this TIP activates, expiring nonce replay state is stored in the TIP-1009 ring layout:

```text
slot 1: expiringNonceSeen[hash] -> expiry
slot 2: expiringNonceRing[index] -> hash
slot 3: expiringNonceRingPtr
```

At the activation block:

1. New expiring nonce insertions MUST use `expiringNonceCells` at slot 4.
2. The old ring pointer, ring mapping, and seen mapping MUST NOT be updated.
3. Slots 1, 2, and 3 MUST remain reserved.
4. Payload-builder prewarming MUST switch to the time-wheel first-cell slot.
5. Transaction pool replay tracking MUST switch to sender-scoped replay hashes and/or time-wheel

>    cell matches instead of detecting inclusion by `expiringNonceSeen[hash]` storage writes.

Reviewer: > MUST switch to sender-scoped replay hashes

Replay detection already uses sender-scoped replay hashes, doesn't it? Afaiu this TIP isn't changing how the replay hash is being calculated.

> MUST switch to sender-scoped replay hashes and/or time-wheel cell matches

This feels ambiguous to me, especially the "and/or" part... replay hashes in the time-wheel cells become the only replay detection mechanism, I'm not sure why there's an "or" in here.

The inclusive slot range
`[EXPIRING_NONCE_CELL_BASE_SLOT, EXPIRING_NONCE_CELL_END_SLOT]` is reserved forever for expiring
nonce replay cells. Future Nonce precompile fields MUST NOT be assigned inside this range.
Mapping-derived slots are assumed collision-resistant under the same Keccak preimage assumptions as
Solidity mapping layout.

Each replay cell stores:

```text
cell = valid_before_u64 || replay_fingerprint_192
```

where:

```text
stored_valid_before = high 64 bits of cell
stored_fingerprint  = low 192 bits of cell
fingerprint         = low 192 bits of H
```

A zero storage word has `stored_valid_before == 0` and is reusable.

Slots 1, 2, and 3 MUST remain reserved forever unless a later hardfork explicitly migrates or
clears the old ring state. Reusing those slots for unrelated fields risks reading stale pre-fork
data.

## Cell Selection

Let:

```text
H = sender-scoped expiring nonce replay hash
V = tx.valid_before
T = block.timestamp
```

`H` MUST be the canonical protocol-computed replay hash, `keccak256(encode_for_signing || sender)`.
It is not user-provided input.

The transaction first chooses its time bucket:

```text
bucket = V % EXPIRING_NONCE_BUCKET_COUNT
```

> It then chooses a deterministic probe path inside that bucket:

Author: This shouldn't be a problem, censoring someone in this way costs money on the attacker's side every 5 minutes. And users can trivially bypass this by using 2d or protocol nonces.

    continue

reject ExpiringNonceSetFull
```

`stored_v != V` is safe to overwrite because the bucket count is larger than the expiry window. Two
different `valid_before` values in the same bucket differ by at least 32 seconds, while expiring
nonce transactions are live for at most 30 seconds.

`ExpiringNonceSetFull` means the transaction's own deterministic probe path is full of different
live fingerprints for the same `valid_before`. Under the time-wheel design this is local
congestion, not global table exhaustion.

## Gas Pricing

The common first-cell path is priced as one cold read and one reset-price write:

```text
EXPIRING_NONCE_BASE_GAS = COLD_SLOAD_COST + WARM_SSTORE_RESET
                        = 2_100 + 2_900
                        = 5_000
```

This charge intentionally does not apply TIP-1000 permanent new-slot pricing to first-touch replay
cells. Expiring nonce cells are bounded temporary protocol state, not user-created permanent state.
The maximum table footprint is capped by `EXPIRING_NONCE_BUCKET_COUNT *
EXPIRING_NONCE_BUCKET_CAPACITY`.

Implementations MUST ensure collision-heavy probe paths cannot create unbounded uncharged work.

> They MAY either charge `COLD_SLOAD_COST` for each additional probed cell beyond the first, or use an

Author: Addressed in https://github.com/tempoxyz/tempo/commit/5c94d8058fe377fdf6c69ac6f873bd5125093f84.

> equivalent dynamic metering mechanism in the precompile storage layer. With the chosen capacity and

Author: Addressed in https://github.com/tempoxyz/tempo/commit/5c94d8058fe377fdf6c69ac6f873bd5125093f84.

particular, 65,536 cells per bucket doubles load to 30.52%, and 32,768 cells per bucket pushes load
above 60%, where probe length and probe-exhaustion risk become much more sensitive to traffic
bursts or adversarial hash selection.

Larger tables reduce collisions but mostly buy marginal latency improvement at the cost of much
larger bounded state. The 16,777,216-cell parameter explored in implementation experiments has
negligible collisions but a 32 GiB slot+value full-table footprint.

## Hardfork Migration

Before this TIP activates, expiring nonce replay state is stored in the TIP-1009 ring layout:

```text
slot 1: expiringNonceSeen[hash] -> expiry
slot 2: expiringNonceRing[index] -> hash
slot 3: expiringNonceRingPtr
```

At the activation block:

1. New expiring nonce insertions MUST use the direct replay-cell slot range
   `[EXPIRING_NONCE_CELL_BASE_SLOT, EXPIRING_NONCE_CELL_END_SLOT]`.
2. The old ring pointer, ring mapping, and seen mapping MUST NOT be updated.
3. Slots 1, 2, and 3 MUST remain reserved.
4. Payload-builder prewarming MUST switch to the time-wheel first-cell slot.
5. Transaction pool deduplication MUST continue to key expiring nonce transactions by `H`.

>    Inclusion tracking MUST stop watching `expiringNonceSeen[H]`; it MUST remove a pooled

Author: Addressed in https://github.com/tempoxyz/tempo/commit/b98a8214780f3029074b3e912b09a69f3c8310b7.

`valid_before` selects one of 301 reusable time buckets, and the sender-scoped transaction replay
hash directly selects a deterministic probe path inside that bucket.

The table is sized at 131,072 cells per bucket, for 39,452,672 total cells. A fully populated table
is 1.18 GiB of cell values, or 2.35 GiB when counted as `(storage slot, storage value)` pairs.

## Motivation

The current expiring nonce ring design keeps replay state bounded, but every accepted transaction
updates a global append structure. The common steady-state path reads and writes several storage
locations: `seen[hash]`, `ring[ptr]`, `seen[old_hash]`, and the ring pointer. The global pointer
also complicates payload-builder prewarming because the storage slot touched by a transaction
depends on how many expiring nonce transactions came before it in the block.

The time wheel keeps replay state bounded without a global append pointer. The common path is a
read and write of one replay-table cell derived from transaction contents, so the builder can
prewarm the first cell directly from `(replay_hash, valid_before)`. Replay cells use a reserved

> direct storage range rather than Solidity mapping layout, so probing a cell does not require a

Reviewer 2: contiguous slots? seems like too much implementation leakage

mapping-slot hash.

Increasing the validity window to 5 minutes gives partners and relayers more flexibility around
submission, batching, and network delay without requiring them to continuously refresh signatures
inside a 30 second window.

## Assumptions

1. Expiring nonce transactions use the TIP-1077 validity window:
   `block.timestamp < valid_before <= block.timestamp + 300`.

> 2. Block timestamps are monotonically nondecreasing across canonical blocks.

Reviewer 2: ```suggestion
2. Block timestamps monotonically increase across canonical blocks.
```

> 3. Replay hashes are uniformly distributed enough that bucket positions are random for operational

Reviewer 2: 3  is very vague

   sizing. Adversaries can grind transaction contents, so probe exhaustion remains a consensus
   error rather than an impossible case.
4. The target load for this parameter set is 20,000 expiring nonce transactions per second,
   sustained over the 5-minute validity window.
5. Traffic is assumed to be roughly constant over the 300 valid `valid_before` seconds. A workload
   that intentionally concentrates all transactions into one exact `valid_before` second is still
   accepted probabilistically, but has higher collision pressure.
6. Expiring nonce replay cells are bounded temporary protocol state. They do not pay TIP-1000
   permanent new-slot pricing on first touch and do not mint or consume TIP-1060 storage credits.

If the target TPS assumption is exceeded for long periods, collision frequency and probe-exhausted
rejections increase. A future TIP can raise `EXPIRING_NONCE_BUCKET_CAPACITY` without changing the
replay algorithm.

## Threat Model

> Users and relayers are untrusted. They may replay old signed transactions, submit many independent

Reviewer 2: ```suggestion
Users and relayers are untrusted. They may attempt to replay old signed transactions, submit many independent
```

expiring nonce transactions with the same `valid_before`, or grind transaction contents to seek
cell collisions.

> Builders and validators are untrusted for transaction ordering but trusted to execute the canonical

Reviewer 2: ```suggestion
Builders and validators are untrusted for transaction ordering but required to execute the canonical
```

> state transition rules. A block containing a replay, an expired expiring nonce transaction, or a

Reviewer 2: ```suggestion
state transition rules. A block containing a replay, an expired expiring-nonce-transaction, or a
```

transaction whose probe path is exhausted is invalid.

---

# Specification

## Constants

```text
PRE_TIP_1077_MAX_EXPIRY_SECS  = 30
EXPIRING_NONCE_MAX_EXPIRY_SECS = 300
EXPIRING_NONCE_BUCKET_COUNT    = 301
EXPIRING_NONCE_BUCKET_CAPACITY = 131_072
EXPIRING_NONCE_POSITION_BITS   = 17
EXPIRING_NONCE_MAX_PROBES      = 5

EXPIRING_NONCE_CELL_COUNT      = 39_452_672
EXPIRING_NONCE_CELL_BASE_SLOT  = 0x387ee7e371ffafdf29537b927a83c9af016eb1493da1fe163fab229fa2400000
EXPIRING_NONCE_CELL_END_SLOT   = 0x387ee7e371ffafdf29537b927a83c9af016eb1493da1fe163fab229fa499ffff
EXPIRING_NONCE_CELL_SLOT_DOMAIN = "tempo.expiringNonceCells.v1"
EXPIRING_NONCE_FINGERPRINT_BITS = 192
```

`EXPIRING_NONCE_MAX_EXPIRY_SECS` increases TIP-1009's maximum expiry from 30 seconds to

> 300 seconds. `PRE_TIP_1077_MAX_EXPIRY_SECS` is used only by the hardfork migration dual-check for
> old ring entries.

Reviewer 2: ```suggestion
300 seconds. `PRE_TIP_1077_MAX_EXPIRY_SECS` is used expiring nonce transactions stored in the nonce ring buffer prior to T8.
```

`EXPIRING_NONCE_BUCKET_COUNT` MUST be greater than `EXPIRING_NONCE_MAX_EXPIRY_SECS`.
`EXPIRING_NONCE_BUCKET_CAPACITY` MUST be a power of two.
`EXPIRING_NONCE_POSITION_BITS` MUST equal `log2(EXPIRING_NONCE_BUCKET_CAPACITY)`.
`EXPIRING_NONCE_POSITION_BITS * EXPIRING_NONCE_MAX_PROBES` MUST be less than or equal to 256.
`EXPIRING_NONCE_CELL_COUNT` MUST equal
`EXPIRING_NONCE_BUCKET_COUNT * EXPIRING_NONCE_BUCKET_CAPACITY`.

`EXPIRING_NONCE_CELL_BASE_SLOT` is the fixed direct-range base derived from
`keccak256(EXPIRING_NONCE_CELL_SLOT_DOMAIN)` by clearing the low 22 bits.
`EXPIRING_NONCE_CELL_END_SLOT` is
`EXPIRING_NONCE_CELL_BASE_SLOT + EXPIRING_NONCE_CELL_COUNT - 1`.
The domain documents how the range was selected; implementations MUST use the fixed constants and

> MUST NOT compute this hash while processing transactions.

Reviewer 2: seems unnecessary to state it doesn't need to be hash and that its slot is directly used

## Storage Layout

The Nonce precompile keeps the existing 2D nonce mapping and reserves a direct replay-cell storage
range:

```solidity
contract Nonce {
    mapping(address => mapping(uint256 => uint64)) public nonces;          // slot 0

    // Slots 1, 2, and 3 are reserved for the pre-TIP-1077 expiring nonce
    // seen mapping, ring mapping, and ring pointer.
}
```

Replay cells are not stored in a Solidity-style mapping. For `cell_id` in
`[0, EXPIRING_NONCE_CELL_COUNT)`, the corresponding storage slot is:

```text
cell_slot = EXPIRING_NONCE_CELL_BASE_SLOT + cell_id
```

The inclusive slot range
`[EXPIRING_NONCE_CELL_BASE_SLOT, EXPIRING_NONCE_CELL_END_SLOT]` is reserved forever for expiring
nonce replay cells. Future Nonce precompile fields MUST NOT be assigned inside this range.
Mapping-derived slots are assumed collision-resistant under the same Keccak preimage assumptions as
Solidity mapping layout.

Each replay cell stores:

```text
cell = valid_before_u64 || replay_fingerprint_192
```

where:

```text
stored_valid_before = high 64 bits of cell
stored_fingerprint  = low 192 bits of cell
fingerprint         = low 192 bits of H
```

> A zero storage word has `stored_valid_before == 0` and is reusable.

Reviewer 2: what does it mean it's reusable?

isnt' this an invalid state?
https://github.com/tempoxyz/tempo/blob/4d1ed10ea27a0ad2f9f4f3dfb0b40447b0fed409/crates/transaction-pool/src/transaction.rs#L517

Slots 1, 2, and 3 MUST remain reserved forever unless a later hardfork explicitly migrates or
clears the old ring state. Reusing those slots for unrelated fields risks reading stale pre-fork
data.

## Cell Selection

Let:

```text
H = sender-scoped expiring nonce replay hash
V = tx.valid_before
T = block.timestamp
```

`H` MUST be the canonical protocol-computed replay hash, `keccak256(encode_for_signing || sender)`.
It is not user-provided input.

> The transaction first chooses its time bucket:

Reviewer 2: ```suggestion
The transaction's time bucket is designated:
```

```text
bucket = V % EXPIRING_NONCE_BUCKET_COUNT
```

It then chooses a deterministic probe path inside that bucket. For probe `i`:

```text
position_i = uint17(H[i * EXPIRING_NONCE_POSITION_BITS
                    .. (i + 1) * EXPIRING_NONCE_POSITION_BITS])
cell_id_i  = bucket * EXPIRING_NONCE_BUCKET_CAPACITY + position_i
slot_i     = EXPIRING_NONCE_CELL_BASE_SLOT + cell_id_i
```

With the chosen constants, the five probe positions are `H[0..17]`, `H[17..34]`, `H[34..51]`,
`H[51..68]`, and `H[68..85]`. Since `EXPIRING_NONCE_BUCKET_CAPACITY` is `2^17`, each 17-bit chunk
selects one cell in the bucket without modulo bias or rejection sampling.

## Replay Algorithm

> An expiring nonce transaction MUST satisfy the TIP-1009 transaction constraints as updated by this

Reviewer 2: "in addition"?

TIP:

```text
tx.nonce_key == TEMPO_EXPIRING_NONCE_KEY
tx.nonce == 0
V is present
T < V <= T + EXPIRING_NONCE_MAX_EXPIRY_SECS
```

Then the Nonce precompile executes:

```text
fingerprint = H[8..32]

for i in 0..EXPIRING_NONCE_MAX_PROBES:
    cell_id = cell_id_i
    slot = EXPIRING_NONCE_CELL_BASE_SLOT + cell_id
    word = sload(slot)
    stored_v = high_64_bits(word)
    stored_f = low_192_bits(word)

    if stored_v != V:
        sstore(slot, V || fingerprint)
        accept

    if stored_f == fingerprint:
        reject ExpiringNonceReplay

    continue

> reject ExpiringNonceSetFull

Reviewer 2: there is not a return here, so this seems to indicate if always fails? or that is what `accept` it?

This charge intentionally does not apply TIP-1000 permanent new-slot pricing to first-touch replay
cells. Expiring nonce cells are bounded temporary protocol state, not user-created permanent state.
The maximum table footprint is capped by `EXPIRING_NONCE_BUCKET_COUNT *
EXPIRING_NONCE_BUCKET_CAPACITY`.

The replay insertion function MUST return the accepted probe count, and the EVM handler MUST charge
the formula above explicitly as regular gas while keeping replay-cell writes exempt from
TIP-1000/TIP-1060 state accounting. With the chosen capacity and the 20k TPS sizing assumption, the
expected accepted transaction reads about 1.085 cells and pays about 5,179 gas for TIP-1077 replay
bookkeeping.

No expiring nonce insertion or later overwrite mints or consumes TIP-1060 storage credits.

The time-wheel lookup performs no Keccak beyond the canonical replay hash `H`. If `H` has already
been computed by transaction validation or pool deduplication, deriving `bucket`, `position_i`,
`fingerprint`, `cell_id`, and `slot` requires only slicing and arithmetic:

```text
steady-state lookup keccaks after H = 0
steady-state lookup keccaks total   = 1, if H must be computed
```

> During the temporary hardfork migration dual-check, implementations still perform the old

Reviewer 2: The ordering is a bit hard to follow as I hit "Parameter Selection" when I expect the context in "Hardfork Migration"

20k TPS.

Smaller tables reduce the maximum footprint but increase average reads and tail risk. In
particular, 65,536 cells per bucket would halve the footprint to 1.18 GiB but double load to
30.52%, and 32,768 cells per bucket pushes load above 60%, where probe length and probe-exhaustion
risk become much more sensitive to traffic bursts or adversarial hash selection.

Larger tables reduce collisions but mostly buy marginal latency improvement at the cost of much
larger bounded state. The 16,777,216-cell parameter explored in implementation experiments has
negligible collisions but a 301 GiB slot+value full-table footprint with a 5-minute window.

## Hardfork Migration

Before this TIP activates, expiring nonce replay state is stored in the TIP-1009 ring layout:

```text
slot 1: expiringNonceSeen[hash] -> expiry
slot 2: expiringNonceRing[index] -> hash
slot 3: expiringNonceRingPtr
```

At the activation block:

1. New expiring nonce insertions MUST use the direct replay-cell slot range
   `[EXPIRING_NONCE_CELL_BASE_SLOT, EXPIRING_NONCE_CELL_END_SLOT]`.
2. The old ring pointer, ring mapping, and seen mapping MUST NOT be updated.
3. Slots 1, 2, and 3 MUST remain reserved.

> 4. Payload-builder prewarming MUST switch to the time-wheel first-cell slot.

Reviewer 2: This isn't a protocol change and elsewhere the "dual-check" would imply some read both

   activation MUST be rejected by the migration dual-check.
3. **Expiry bounds.** Transactions with `valid_before <= block.timestamp` or
   `valid_before > block.timestamp + 300` MUST be rejected.
4. **Bucket expiry proof.** Within a bucket, `stored_valid_before != tx.valid_before` MUST imply the
   stored record is not live.
5. **Collision handling.** A live collision with a different fingerprint MUST probe the next
   deterministic cell, up to `EXPIRING_NONCE_MAX_PROBES`.
6. **Probe exhaustion.** If all probed cells contain different live fingerprints for the same
   `valid_before`, the transaction MUST be rejected with `ExpiringNonceSetFull`.
7. **Bounded state.** New replay state MUST be limited to
   `EXPIRING_NONCE_BUCKET_COUNT * EXPIRING_NONCE_BUCKET_CAPACITY` cells in the reserved direct slot
   range.
8. **No explicit cleanup dependency.** Replay correctness MUST NOT depend on offchain pruning,
   MEV services, or TIP-1060 storage credits.
9. **No nonce mutation.** Expiring nonce transactions MUST NOT increment protocol nonces or 2D
   nonces.
10. **Storage layout hygiene.** Old ring slots 1, 2, and 3 MUST remain reserved after activation,

>     and future Nonce precompile fields MUST NOT overlap the direct replay-cell slot range.

Reviewer 2: ```suggestion
8. **No nonce mutation.** Expiring nonce transactions MUST NOT increment protocol nonces or 2D
   nonces.
9. **Storage layout compatbility.** Deprecated ring buffer slots, 1, 2, and 3, MUST remain reserved after activation,
    and future Nonce precompile fields MUST NOT overlap the direct replay-cell slot range.
```
Not really an invariant

that transaction's derived 64-bit fingerprint in one of the three packed slots, not merely because a
watched cell became non-zero.

To preserve replay protection across the hardfork boundary, implementations MUST perform a
temporary dual replay check:

```text
activation_timestamp = timestamp of the first block executing TIP-1077
old_seen_cutoff      = activation_timestamp + PRE_TIP_1077_MAX_EXPIRY_SECS

if block.timestamp <= old_seen_cutoff:
    reject if old_expiringNonceSeen[H] > block.timestamp

always:
    check and mark the TIP-1077 time-wheel cell
```

The old ring only contains transactions accepted under TIP-1009's 30 second maximum expiry, so the
dual-check cutoff uses `PRE_TIP_1077_MAX_EXPIRY_SECS` rather than the new 300 second window. After
`old_seen_cutoff`, every transaction recorded only in the old ring layout must have expired, so the
old seen mapping is no longer needed for replay validation.

The hardfork does not need to enumerate or delete old ring state. The old slots become orphaned
reserved state. A later state-cleanup hardfork MAY clear them, but replay safety does not depend on
that cleanup.

> ## Alternatives Considered

Reviewer 2: Would be nice to discuss status cache and have an idea of its footprint versus 1077
